template renderer improvements

chassing opened the pull request:
This PR has several `template-renderer` improvements and bugfixes:

###  **:sparkles: render template.patch.path**

`template.patch.path` can be a template, too.


### **:bug: template-renderer: fix --clone-repo option**

`-clone-repo` command line option is a flag

### **:bug: template-renderer: strip targetPath**

Strip whitespaces and new lines from `targetPath`. Especially useful if `targetPath` is multiline and a template. 

### **:bug: template-renderer: update file cache for after each template**

In case two templates are dependent, e.g., the 1st one introduces an `openshiftResource` for an AWS account, and the 2nd adds an additional resource, then the internal file cache must have the result from the 1st one. See [this as an example](https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/109948/diffs#e159c5d0c1477875efa58ad0e33e050a7fa84e98_444_416). `aws-resource-exporter` introduces `provisioner: $ref: /aws/jeanluc/account.yml` and `cloudwatch-exporter` adds another `resource.provider: aws-iam-service-account`

### **:bug: gitlab: fix Calling a `list()` method without specifying `get_all=True` or `iterator=True` will return a maximum of 20 items warning**

python-gitlab 4.6 upgrade bug

patjlm commented on reconcile/templating/lib/rendering.py:
```diff
@@ -78,7 +78,7 @@ def render_output(self) -> str:
 
     def render_target_path(self) -> str:
-        return self._render_template(self.template.target_path)
+        return self._render_template(self.template.target_path).strip().strip(r"\n")
```
don't you need an other `.strip()` at the end if new line and spaces are mixed?

hemslo replied:
we need to leave one newline at EOF, will this strip everything?

chassing replied:
We want to strip anything because it's a file path string.

hemslo replied:
I think `\n` is also removed in `strip()`, any edge cases?

chassing replied:
jupp, you're right, one `strip()` is enough. fixed in [3a2ec52](https://github.com/app-sre/qontract-reconcile/pull/4499/commits/3a2ec520b8e97e74ddf01189d537552c9a323610)

patjlm commented on reconcile/test/fixtures/templating/full.yaml:
```diff
@@ -1,7 +1,9 @@
-  targetPath: /{{bar}}/foo.yml
+  # test targetPath with templating, multiline and whitespaces
+  targetPath: |
+    {% if true %} /{{bar}}/foo.yml {% endif %}
```
maybe
```
{% if true %}
      /{{bar}}/foo.yml     # with ending spaces
{% endif %}
```

chassing replied:
applied in [6916c35](https://github.com/app-sre/qontract-reconcile/pull/4499/commits/6916c354fdd911569dc41eb58d35604a8074905d)

hemslo commented on reconcile/utils/gitlab_api.py:
```diff
@@ -130,7 +130,7 @@ def __init__(
     def project_main_branch(self) -> str:
         return next(
-            (b.name for b in self.project.branches.list() if b.default),
+            (b.name for b in self.project.branches.list(get_all=True) if b.default),
```
use `iterator=True` may save some requests if default branch returns in early pages

chassing replied:
Applied in [0bc805f](https://github.com/app-sre/qontract-reconcile/pull/4499/commits/0bc805ff02dd78b3980ac90cc8418dd0652bf972)

hemslo commented on reconcile/templating/renderer.py:
```diff
@@ -303,27 +299,22 @@ def reconcile(
-                            input,
-                            item,
-                        )
+                    self.reconcile_template_collection(
+                        collection=c,
+                        gql_api=gql_no_validation,
+                        persistence=p,
+                        ruamel_instance=ruamel_instance,
+                        input=template_input,
+                        each=item,
```
output is still needed, see pr https://github.com/app-sre/qontract-reconcile/pull/4488

chassing commented on reconcile/templating/renderer.py:
```diff
@@ -303,27 +299,22 @@ def reconcile(
+                        input=template_input,
+                        each=item,
                     )
-                if not dry_run and outputs:
```
This dry run handling isn't needed because it's already covered in the `PersistenceTransaction.__exit__` method.

hemslo reviewed (approved):
LGTM!

chassing commented:
@BumbleFeng, I rebased this PR and refactored your `TemplateResult` changes slightly. The idea is that only `ClonedRepoGitlabPersistence` needs it, and all other `FilePersistence` still operate on `TemplateOutput`.

chassing commented on reconcile/templating/renderer.py:
```diff
@@ -129,39 +140,51 @@ def __enter__(self) -> Self:
-            auto_approved = [o for o in result.outputs if o.auto_approved]
+        self.result.outputs = self.outputs
+        if self.result.enable_auto_approval:
+            auto_approved = [o for o in self.result.outputs if o.auto_approved]
```
TBH, it looks like there is a bug. The current code, for template collections with `auto_approved`, will create just one MR with the `auto_approved` outputs, and all others are skipped!? :(

@janboll, wdyt?

chassing replied:
I've tested it, and yes, it's a bug :( It creates just one MR with all auto-approved outputs, and everything else is skipped.

chassing replied:
Fixed in [7276db7](https://github.com/app-sre/qontract-reconcile/pull/4499/commits/7276db751ca38457368c36f833e770dc79254110)

janboll replied:
This was Intentional. Create one MR with only auto_approved changes, and after that is merged create another one. The idea was to converge as fastest as possible. 
The idea was to have only one MR per collection to reduce the number of open MRs/merge conflicts
